Feature/storage s3

OmerGronich opened the pull request.

drorrose commented:
@davidmeirlevy @OmerGronich 
On my computer couldn't run the service due to error (not finding aws-sdk, even though locally installed and in package.json). Does it reproduce on any of your computers?

OmerGronich commented:
@davidmeirlevy @drorrose @or-yam 
Renaming asssets is still WIP.

OmerGronich commented:
@davidmeirlevy  
in gcs services, when you load files you return the type property in the return value. 
gcs has the metadata.kind built in to the api but could not find a way to do this with s3.

`
async function loadFiles(storage, identifier = '/') {
  const gcs = new Gcs(storage)
  const fullPath = path.join(storage.metadata.basePath || '/', identifier)

  let list
  try {
    await gcs.ready
    list = await gcs.list(fullPath)
  } catch (e) {
    throw new Error(e.message || 'failed to get list of assets from: ' + fullPath)
  } finally {
    // run on background
    // TODO: reuse storage connection
    gcs.destroy()
  }

  return list.map((asset) => {
    const fileIdentifier = path.join(identifier, asset.metadata.name)

    return {
      name: asset.metadata.name,
      identifier: fileIdentifier,
      updated: asset.metadata.updated,
      **type: getAssetType(asset.metadata)**,
      publicUrl: joinUrl(storage.metadata.publicUrl, fileIdentifier)
    }
  })
}
`

davidmeirlevy commented:
@OmerGronich rach provider use a different approach to distinguish the file types, but as an assets-service, we should at least tell the user is this is an image, a folder, or another "asset".

OmerGronich commented:
> @OmerGronich rach provider use a different approach to distinguish the file types, but as an assets-service, we should at least tell the user is this is an image, a folder, or another "asset".

Got it 👍
I will figure it out.

davidmeirlevy commented on server/models/s3.js:
```diff
@@ -0,0 +1,115 @@
+      (decrypted) => {
+        const { accessKeyId, secretAccessKey } = decrypted.value;
+
+        AWS.config.setPromisesDependency();
```
This service is stateless and can run multiple times.. maybe this line should run only once?

OmerGronich replied:
👍 no problem

moved outside of class -> [commit 277cb0895d8c838429c6a97533fbbb663b723c0c](https://github.com/greenpress/assets-service/pull/17/commits/277cb0895d8c838429c6a97533fbbb663b723c0c)

davidmeirlevy commented on server/models/s3.js:
```diff
@@ -0,0 +1,115 @@
+  }
+
+  async destroy() {
+
```
Should destroy the this._client connection.

OmerGronich replied:
What does destroying the connection mean? 

All this._client requests are stateless, I'm pretty sure there is no continuous connection to S3. 
Can I delete this function?

davidmeirlevy commented on server/models/s3.js:
```diff
@@ -0,0 +1,120 @@
+        }
+      }));
+
+      return [...files, ...folders];
```
it's common to put folders first.

OmerGronich replied:
done

davidmeirlevy commented on server/models/s3.js:
```diff
@@ -0,0 +1,120 @@
+        Bucket: this.bucket.name
+      };
+
+      const files = await Promise.all(listedObjects.Contents.map(async content => {
```
Is there no faster way to load basic data for the files?
there are a lot of requests here.

maybe the next phase is to start with the caching mechanism here.

OmerGronich replied:
No problem! The code is no longer making the extra request to s3 to get the object metadata. Now deriving the asset type  with our getAssetType function.

davidmeirlevy commented on server/models/s3.js:
```diff
@@ -0,0 +1,120 @@
+        Delete: { Objects: [] }
+      };
+
+      listedObjects.Contents.forEach(({ Key }) => {
```
why not to use map, like:
```
      const deleteParams = {
        Bucket: this.bucket.name,
        Delete: { Objects: listedObjects.Contents.map({Key} => ({Key})) }
      };
```

OmerGronich replied:
Cool :) 
Refactored👍

davidmeirlevy commented on server/models/s3.js:
```diff
@@ -0,0 +1,120 @@
+
+      await this._client.deleteObjects(deleteParams).promise();
+
+      if (listedObjects.IsTruncated) await this.remove(file);
```
this looks a bit recursive. why is that?

OmerGronich replied:
the isTruncated boolean is s3's pagination flag. True = there are still assets left to retrieve

davidmeirlevy commented on server/routes/assets.js:
```diff
@@ -1,14 +1,20 @@
 
-function empty (_, res) {
-  return res.status(200).json({ message: 'endpoint not yet exists' }).end()
+function empty(_, res) {
```
you can remove this function we don't use it anymore.

OmerGronich replied:
done

davidmeirlevy reviewed (commented):
minor changes left. this can be approved, but it can be even better with those changes. :)
